### PR TITLE
Add C# support for dynamic server SSL certificate reloading

### PR DESCRIPTION
**Pull request by ak800i**
Implemented API integration for facilitating dynamic server SSL certificate reloading in C# via callback (config fetcher).

@markdroth
@jtattermusch

### CODE REVIEW

**Comment by ak800i** [comment]
Related issue: https://github.com/grpc/grpc/issues/16118

**Comment by markdroth** [comment]
Note that we are in the process of changing the C-core API for credential reloading, which is currently still experimental.  We should probably make sure that this change does not conflict with that work -- i.e., this PR should only be merged if we will not require any C# API changes when the C-core API change is complete.

Also, note that if this is going to be a non-experimental API in C#, it will require a gRFC.

File: src/csharp/Grpc.Core/Internal/ServerCertificateConfigCallbackRegistration.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+                ServerCertificateConfig serverCertificateConfig = this.serverCertificateConfigCallback();
+
+                // Write the pointer to the certificate to the location where arg1 is pointing to:
+                ServerCertificateConfigSafeHandle nativeServerCertificateConfig = serverCertificateConfig.ToNative();
+                Native.grpcsharp_write_ssl_server_certificate_config_to_pointer(arg1, nativeServerCertificateConfig);
+
+                // Perhaps the pointer arithmetic should be done in Native should be done in CLR:
+                    //GCHandle handle = GCHandle.Alloc(nativeServerCertificateConfig);
+                    //IntPtr pointerToNativeHandle = GCHandle.ToIntPtr(handle);
+                    //Marshal.WriteIntPtr(arg1, pointerToNativeHandle);
```
Comment (ak800i) [inline src/csharp/Grpc.Core/Internal/ServerCertificateConfigCallbackRegistration.cs:61]: Please take a look here and suggest whether the approach I took (with grpcsharp_write_ssl_server_certificate_config_to_pointer) is satisfying.
I have tried doing the pointer arithmetic here but was unsuccessful.

File: src/csharp/Grpc.Core/Internal/ServerCertificateConfigCallbackRegistration.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+                // We could somehow retrieve the old cert config and compare instead of passing .New unconditinally.
+                return (int)SslCertificateConfigReloadStatus.New;
```
Comment (ak800i) [inline src/csharp/Grpc.Core/Internal/ServerCertificateConfigCallbackRegistration.cs:64]: Suggestions on how to retrieve the old cert are welcome.

File: src/csharp/Grpc.Core/Internal/ServerCredentialsSafeHandle.cs [hunk 3]
```diff
@@ -32,13 +32,29 @@ private ServerCredentialsSafeHandle()
...
-        public static ServerCredentialsSafeHandle CreateSslCredentials(string pemRootCerts, string[] keyCertPairCertChainArray, string[] keyCertPairPrivateKeyArray, SslClientCertificateRequestType clientCertificateRequest)
+        public static ServerCredentialsSafeHandle CreateSslCredentials(
+            string pemRootCerts,
+            string[] keyCertPairCertChainArray,
+            string[] keyCertPairPrivateKeyArray,
+            SslClientCertificateRequestType clientCertificateRequest,
+            IntPtr serverCertificateConfigCallbackTag,
+            IntPtr userData)
         {
-            GrpcPreconditions.CheckArgument(keyCertPairCertChainArray.Length == keyCertPairPrivateKeyArray.Length);
```
Comment (ak800i) [inline src/csharp/Grpc.Core/Internal/ServerCredentialsSafeHandle.cs:37]: Note to self: This line should not have been deleted. Restore it.

File: src/csharp/ext/grpc_csharp_ext.c [hunk 4]
```diff
@@ -1076,6 +1076,82 @@ grpcsharp_ssl_server_credentials_create(
...
+
+GPR_EXPORT grpc_ssl_server_credentials_options* GPR_CALLTYPE
+grpcsharp_ssl_server_credentials_create_options_using_config(
+    grpc_ssl_client_certificate_request_type client_certificate_request,
+    grpc_ssl_server_certificate_config* certificate_config) {
+  return grpc_ssl_server_credentials_create_options_using_config(
+      client_certificate_request, certificate_config);
+}
+
+grpc_ssl_server_certificate_config_callback saved_cb = NULL;
```
Comment (ak800i) [inline src/csharp/ext/grpc_csharp_ext.c:1117]: Is this the proper way to save the callback tag?

**Comment by ak800i** [comment]
@jtattermusch @jiangtaoli2016 just a kind reminder on this

**Comment by ak800i** [comment]
@jtattermusch @jiangtaoli2016
Is there an ETA for reviewing this? What are some average times from PR being submitted to the the start of reviewing?
I am a Microsoft employee and this feature is needed by our product. We are blocked on this.
I understand that you might be busy.
I am just wondering if there is a problem with this PR (i.e. whether if fell through some cracks) and fear that it is not even be queued for reviewing. If you are busy, that is understandable, I would just like info that this PR is eventually going to get reviewed.

I fear that you might filter PRs by their EasyCLA status, and that PRs whose EasyCLA is not passing are filtered out.
The commit (9dd14c1) is missing the User's ID, because I omitted my email I have since edited the commit and force-pushed it, overwriting the commit 9dd14c1 with b44593c but EasyCLA did not pick this up. I will try to push another commit in the hopes that EasyCLA will get back on track.

**Comment by jwfx** [comment]
I'm very interested in this PR, as support for this feature is missing in the C# part for quite some time now and it actually affects design decisions for gRPC server integrations. (Restarting servers for changing credentials is a no-go..)

Is there any chance this could be moved up in the review queue?

**Comment by jtattermusch** [comment]
@ak800i  sorry for delay.

I can review from the C# integration perspective, I'm not too knowledgeable about the new C-core credentials API. Is the API documented somewhere (besides the original PR https://github.com/grpc/grpc/pull/12644)? An example for using the API would also be useful.

@yihuazhang  @jiangtaoli2016:
- is the API being wrapped by this PR the right API to use? (and also the one we're supposed to be using for this purpose in the long run? I'm not sure about the relationship between the current API and the new TLS credentials from https://github.com/grpc/proposal/pull/98/files).
- can you review that API is being used correctly in this PR.

After the answer is yes to both questions, I can look more in detail at the way the wrapping is done in C# (I think the general approach is roughly ok, and the boilerplate I'd expect seems to be there, but I'd need to understand the usage patterns of the credentials refresh API a bit more to be able to review in detail).

**Comment by ak800i** [comment]
@jtattermusch I do not know that any documentation exists.
The goal of this PR is to bring the C# implementation to Go/Python's level.
Here is the the PR that brought this API to Python https://github.com/grpc/grpc/pull/13188
and here is the "official" issue with C# https://github.com/grpc/grpc/issues/16118

**Comment by jwfx** [comment]
@jtattermusch @jiangtaoli2016 

What is the current state of the review?

**Comment by jtattermusch** [comment]
> @jtattermusch @jiangtaoli2016
> 
> What is the current state of the review?

There is an internal thread with the PR author where we're trying to figure out what are the next best steps in terms of supporting the upcoming TLS credentials API described here: https://github.com/grpc/proposal/pull/98/files

**Comment by jwfx** [comment]
> There is an internal thread with the PR author where we're trying to figure out what are the next best steps in terms of supporting the upcoming TLS credentials API described here: https://github.com/grpc/proposal/pull/98/files

Isn't that entirely different conversation?

From my understanding this PR is supposed to close a gap between the C API and the C# implementation that exists since Sept. 2017 when the cert reloading feature was merged.

Would be sad to see this PR going to waste and C# users have to wait another extended period of time for the new API to be supported.

**Comment by jtattermusch** [comment]
> > There is an internal thread with the PR author where we're trying to figure out what are the next best steps in terms of supporting the upcoming TLS credentials API described here: https://github.com/grpc/proposal/pull/98/files
> 
> Isn't that an entirely different conversation?

I don't think it is. We know that a new credentials API is going to be introduced in C-core soon (and the older one is going away), so I want to only add a C# API that won't need to be revisited (modified/removed) in near future.

> 
> From my understanding this PR is supposed to close a gap between the C API and the C# implementation that exists since Sept. 2017 when the cert reloading feature was merged.
> 
> Would be sad to see this PR going to waste and C# users have to wait another extended period of time for the new API to be supported.

I appreciate the effort to close the feature gap (sorry for the delay when reviewing), but I want to make sure it's done the right way. We need to make sure that if we bring in a new C# API, it's going to be compatible with the new C core API that's coming. If we add an API that we know will break soon, it's not going to be useful for anyone.
Let's make sure the proposed API is compatible with https://github.com/grpc/proposal/pull/98/files first.

**Comment by jtattermusch** [comment]
> @jtattermusch who is going to write this short proposal that summarizes the C# public API we want to use going forward?

It would have to be someone from the community (feel free to nominate yourself) - while I believe this would be a useful addition, I unfortunately don't have the cycles to write the proposal myself (as I pointed out earlier).

**Pull request closed without merging**